src/clientdb/Runner.cc

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+/* DEBUG: section 77    Client Database */
+
+#include "squid.h"
+#include "base/RunnersRegistry.h"
+#include "clientdb/Cache.h"
+#include "mgr/Registration.h"
+#include "SquidConfig.h"
+
+namespace ClientDb
+{
+
+/// hooks the ClientDb into Squid's Runner API
+class Runner: public RegisteredRunner
+{
+public:
+    /* RegisteredRunner API */
+    virtual void useConfig() override {
+        if (Config.onoff.client_db)
+            Mgr::RegisterAction("client_list", "Cache Client List", ClientDb::Report, 0, 1);
+    }
+    virtual void finishShutdown() override {
+        ClientDb::Cache.clear();
+    }
```

Comment:
I expect this runner to grow well beyond mgr reporting as more of the clientdb related functionality is updated. So will be leaving it as-is for now.

Reply:
Even as we add clientdb functionality, this runner is likely to remain trivial. The runner will simply call whatever non-trivial code is needed to implement that additional functionality. A runner is just a collection of event hooks. In most cases, the event handling functionality is trivial or implemented elsewhere. Isolating feature X runners into runner-dedicated files makes things worse because it forces otherwise internal X code to be exposed to X users that should not see it.

In the ideal world, these runners would not exist at all -- we would just annotate some functions so that they get automatically registered with the event system by the compiler. We are implementing the same basic event hookup idea using C++ tools at our disposal. Placing these simple wrappers into wrapper-dedicated files makes things worse.